src/mobu/models/jupyter.py

```diff
@@ -1,62 +1,161 @@
 
-from enum import Enum
-from typing import Dict, Optional
+from enum import Enum, auto
+from typing import Dict, List, Optional
 
 from pydantic import BaseModel, Field, validator
 
-__all__ = ["JupyterConfig", "JupyterImage", "JupyterImageClass"]
+__all__ = [
+    "JupyterConfig",
+    "JupyterImageClass",
+    "ControllerImage",
+    "ControllerImages",
+]
 
 
-class JupyterImageClass(Enum):
-    """Possible ways of selecting an image."""
+def to_camel_case(string: str) -> str:
+    """Convert a string to camel case.
+
+    Originally written for use with Pydantic as an alias generator so that the
+    model can be initialized from camel-case input (such as Kubernetes
+    objects).
+
+    Parameters
+    ----------
+    string
+        Input string
+
+    Returns
+    -------
+    str
+        String converted to camel-case with the first character in lowercase.
+    """
+    components = string.split("_")
+    return components[0] + "".join(c.title() for c in components[1:])
+
+
+class CamelCaseModel(BaseModel):
+    """This is what we will use in place of BaseModel for the Spawner
+    Pydantic models.  Any configuration can be given in Helm-appropriate
+    camelCase, but internal Python methods and objects will all be snake_case.
+
+    This isn't actually all that useful here, but since these models are
+    copied from jupyterlabcontroller, which *does* make use of these features,
+    it's easier than messing with the models.
+    """
+
+    class Config:
+        """Pydantic configuration."""
+
+        alias_generator = to_camel_case
+        allow_population_by_field_name = True
+
+
+# Dashify is needed to turn, e.g. "latest_weekly" into the required
+# "latest-weekly" per sqr-066.  It's also handy for SpawnerEnum.
+
+
+def dashify(item: str) -> str:
+    return item.replace("_", "-")
+
+
+class SpawnerEnum(str, Enum):
+    """This will validate that the name is entirely upper case, and
+    will produce auto() values in lower case with underscores turned to
+    dashes.
+    """
 
-    RECOMMENDED = "recommended"
-    LATEST_WEEKLY = "latest-weekly"
-    BY_REFERENCE = "by-reference"
+    def _generate_next_value_(  # type: ignore
+        name, start, count, last_values
+    ) -> str:
+        if name != name.upper():
+            raise RuntimeError("Enum names must be entirely upper-case")
+        return dashify(name.lower())
 
 
-class JupyterImage(BaseModel):
-    """Represents an image to spawn as a Jupyter Lab."""
+class NubladoEnum(str, Enum):
+    """This will validate that the name is entirely upper case, and
+    will produce auto() values in lower case.  This is exactly StrEnum from
+    Python 3.11, except for the validation step."""
 
-    reference: str = Field(
+    def _generate_next_value_(  # type: ignore
+        name, start, count, last_values
+    ) -> str:
+        if name != name.upper():
+            raise RuntimeError("Enum names must be entirely upper-case")
+        return name.lower()
```

Comment:
I don't understand these changes. They seem like a ton of unnecessary complexity. Why can't we just use simple enums?

Reply:
Same as above: I tried to minimize the changes when pulling this over from JupyterLab Controller.  We *do* care about the actual string values there.  I think the right way to go may be to go to Python 3.11 and use its new StrEnum class directly, though.  That doesn't get us the keys-are-uppercase validator, but, meh.